improved computation of interactive marker size

rhaschke opened the pull request:
- use parent_link if group == parent_group (instead of default)
- scale smaller than 5cm is clipped to 5cm (instead of falling back to default)
  If there is a really small end-effector group / link, we should rely on the computation too, shouldn't we?
- clarified size computation, using diameter of AABB

rhaschke committed:
improved computation of interactive marker size

- use parent_link if group == parent_group
- scale smaller than 5cm is clipped to 5cm instead of using default
- clarified size computation, using diameter of AABB

davetcoleman commented on robot_interaction/src/robot_interaction.cpp:
```diff
@@ -116,9 +116,28 @@ void RobotInteraction::addActiveComponent(const InteractiveMarkerConstructorFn &
+  double s = 1.01 * ext.norm();
+
+  // clip scale to 5cm
+  return std::max(0.05, s);
```
how did you choose 5cm here? i've seen some robots whose model does not properly follow metric sizing, i.e. they could be a couple of orders of magnitude larger or smaller than real life meters. Seems this could be error prone hard coding 0.05

rhaschke replied:
I would agree, that the code should simply rely on its measurements, regardless of magnitude.
However, I didn't invent the 5cm limit. It was there before (see below) and I didn't want to break existing code.
Probably it was there for some reason.

davetcoleman replied:
ah indeed! seems like that should have at least be a `static const`

davetcoleman commented on robot_interaction/src/robot_interaction.cpp:
```diff
@@ -320,7 +336,7 @@ void RobotInteraction::decideActiveEndEffectors(const std::string &group, Intera
     // if we have a separate group for the eef, we compute the scale based on
     // it; otherwise, we use a default scale
```
This comment should be updated to reflect your changes

rhaschke committed:
reworked computeLinkMarkerSize()

compute size such that the marker sphere will cover
- a spherical link geometry -> AABB.maxCoeff
- a cubical link geometry -> AABB.norm
-> use average of both values

Virtual links (without any shape) will have a size of AABB of zero dims.
In this case use the dimensions of the closest parent link instead.

rhaschke commented:
added the proposed static const variable
modified the comment as suggested

davetcoleman commented:
This makes sense to me, but I am not able to test it (away from office) so am hesitant to merge. Not sure how to verify this won't cause unexpected visual change for others.

rhaschke commented:
There is no need to hurry. I'm at IROS conference this week too. Of course, this PR will have some visual effects. 
For us, having a Shadow robot hand with five rather small fingertips, it was kind of necessary to have both, a large end-effector marker at the palm (to pull the arm) and smaller end-effectors for the fingertips. With the old code, the marker size fall back to a rather large default size not suitable for fingers. However, the size setting in the rviz gui is global a resize also the palm marker that needs to be larger...

sachinchitta commented:
I just tried this - the markers are much smaller than they used to be. You can try it with the same robot - https://github.com/sachinchitta/moveit_examples

rhaschke commented:
@sachinchitta  Indeed your example robot has a virtual end-effector link (resulting in size zero), which I didn't considered during implementation yet. To handle this case gracefully (instead of returning the `DEFAULT_SIZE`), I now use the size of the closest non-virtual parent link.
This will give a reasonable marker size in that case too.

sachinchitta commented:
This looks better now but would be good to have the size still be a little bigger - I tried it with 3.0 \* size and it was easier to interact with.

rhaschke committed:
further adapted marker size computation

- drop largest extension dimension (-> use cross-section size of elongated link)
- for an end-effector group, consider the sizes of individual links
  instead of the overall size of all links (which becomes huge very fast)
- enlarge marker size by factor of 1.5 when there is only a single eef marker

sachinchitta commented:
LGTM.
